Adding checks for broken bottleneck files

rizasif opened the pull request:
An exception is raised when a broken cached bottleneck files is read. Such a file can be created as a result of sudden system failure. 

This commit refers the issue #2296 (https://github.com/tensorflow/tensorflow/issues/2296). Although closed but the problem still persists. Please view comments by @rizasif92 at the end. The changes contain nothing but error handling.

rizasif committed:
Adding checks for broken bottleneck files

An exception is raised when a broken cached bottleneck files is read. Such a file can be created as a result of sudden system failure. This commit refers the issue #2296

tensorflow-jenkins commented:
Can one of the admins verify this patch?

rizasif commented:
I signed it!

petewarden reviewed (changes requested):
Thanks @rizasif , I really appreciate you putting this fix together! Could I request a slightly different approach though? I would prefer to keep the get_or_create_bottleneck() function always returning a bottleneck.

To ensure that, we could move the section nested inside the `if not os.path.exists(bottleneck_path):` condition into a new function, `create_bottleneck_file`. Then we can call it from that condition, but also have a flag that catches float exceptions when reading and calls it in that case too. Here's some untested code to try and show what I mean:

```
  if not os.path.exists(bottleneck_path):
    create_bottleneck_file(...)
  did_hit_error = False
  try:
    bottleneck_values = [float(x) for x in bottleneck_string.split(',')]
  except:
    print("Invalid float found, recreating bottleneck")
    did_hit_error = True
  if did_hit_error:
    create_bottleneck_file(...)
    # Allow exceptions to propagate here, since they shouldn't happen after a fresh creation
    bottleneck_values = [float(x) for x in bottleneck_string.split(',')]
```
I'm hoping with this approach none of the other changes outside this function should be needed since the contract to always return a list of bottleneck values is preserved.

Does that make sense?

rizasif committed:
handling broken bottleneck files by recreating

The patch detects the broken bottleneck files (might be created as a result of sudden system shut down) and recreates them if required.

rizasif commented:
@petewarden I agree with your solution and was thinking to implement it. I have made the changes as suggested, and tested them as well; I am attaching a screenshot of the output. Kindly review and let me know if anything else is required.
Thank you :)

![capture](https://cloud.githubusercontent.com/assets/8410351/22466834/dae8d6f2-e7e3-11e6-978c-6ebe8fc2a617.PNG)

petewarden commented on tensorflow/examples/image_retraining/retrain.py:
```diff
@@ -481,7 +498,6 @@ def get_random_cached_bottlenecks(sess, image_lists, how_many, category,
       ground_truth[label_index] = 1.0
       bottlenecks.append(bottleneck)
       ground_truths.append(ground_truth)
-      filenames.append(image_name)
```
I think we should put this line back?

petewarden commented on tensorflow/examples/image_retraining/retrain.py:
```diff
@@ -493,11 +509,12 @@ def get_random_cached_bottlenecks(sess, image_lists, how_many, category,
                                               image_index, image_dir, category,
                                               bottleneck_dir, jpeg_data_tensor,
                                               bottleneck_tensor)
-        ground_truth = np.zeros(class_count, dtype=np.float32)
```
Can we revert this change, since now the bottlenecks will never be None?

petewarden reviewed (changes requested):
This looking great, thanks for the updates! I have a couple of minor comments, but once those are addressed this looks good to go in.

rizasif committed:
minor changes for broken bottleneck handling

Reverting some part of code for bottleneck handling changes

rizasif commented:
@petewarden done

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
@rizasif it looks like you added some 4 character indents, which breaks the sanity check. TF uses 2 character indents. Please fix.

rizasif committed:
Resolved indentation errors

This commit refers to the issue #2296

rizasif commented:
Indentation changes made, kindly re-run the tests. Thanks :)

rmlarsen commented:
@tensorflow-jenkins test this please

rmlarsen commented:
Thanks @rizasif. @petewarden can you take another look? If the logics right to you, approve and I'll get it merged.

petewarden reviewed (approved):
This looks great, thanks!

rizasif commented:
Did the tests pass? :/

rmlarsen commented:
@rizasif it looks like you still have some bad indents, e.g.

FAIL: Found 1 non-whitelited pylint errors:
tensorflow/examples/image_retraining/retrain.py:353: [E0001(syntax-error), ] unexpected indent

rizasif committed:
Indentation issue for broken bottleneck script

This commit refers to pull request #7131 and issue #2296

rizasif commented:
@rmlarsen the code seems to be working at my end; so we might have to do this a few more times to resolve these indentation issues. I hope this commit works fine. @tensorflow-jenkins please test :)

rmlarsen commented:
@rizasif Thanks :) 

@tensorflow-jenkins test this please

rmlarsen commented:
@rizasif thanks for the contribution!

rizasif commented:
Pleasure's all mine @rmlarsen 
Thank you so much for all the support. Hope to see you guys again. Take Care :)
